### PR TITLE
Initial checkin

### PR DESCRIPTION
**Pull request by tomtupycape**
App.config
CapeDBClasses.cs
CapeSftpManager.cs
DBConfigHelper.cs
FreyrManager.cs
MDFManager.cs
Program.cs

### CODE REVIEW

File: CapeMSSQLFreyrLink/Program.cs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+            // 6.1. Open MDF File
+            mdf_mgr.openFile(file_to_process_path);
+            // Get columns from MDF
+            List<CapeTableColumn> mdf_columns = mdf_mgr.getTableColumns(ConfigurationManager.AppSettings["mdfTableName"]);
+
+            // Try-Catch used for PostgreSQL connection handling
+            try {
+                // 6.2. Connect to Freyr
+                FreyrManager freyr_mgr = new FreyrManager();
```
Comment (WafflesNotStacks) [inline CapeMSSQLFreyrLink/Program.cs]: suggest breaking this out into its own function if it starts getting any larger

File: CapeMSSQLFreyrLink/Program.cs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                CapeSftpManager sftp_mgr = new CapeSftpManager();
+                sftp_mgr.connect();
+
+                // 2. Update File Cache for all files
+                foreach (SftpFile file in sftp_mgr.getMDFFiles()) {
+                    mdf_mgr.updateFileInCache(file);
+                }
+
+                // 3. Get the first new file
+                KeyValuePair<string, CapeSftpFile> file_cache_entry = mdf_mgr.getFirstNewFileFromCache();
```
Comment (WafflesNotStacks) [inline CapeMSSQLFreyrLink/Program.cs]: why are you only getting the first new MDF file? What about if there are 2+ new files in the list?

Comment (tomtupycape) [inline CapeMSSQLFreyrLink/Program.cs]: The next iteration of the loop after processing will pick it up

File: CapeMSSQLFreyrLink/FreyrManager.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                table_columns_string_list.Add(ctc.Name.ToLower() + " " + MSSQL_TO_POSTGRES_DATA_TYPE_MAP[ctc.Type]);
+            }
+            table_create_sql += string.Join(",", table_columns_string_list);
+            table_create_sql += ") WITH (OIDS=FALSE); ALTER TABLE " + ConfigurationManager.AppSettings["freyrClientsSchema"] + "." + table_name + " OWNER TO " + ConfigurationManager.AppSettings["freyrUsername"];
+            Console.WriteLine("Table Create SQL: " + table_create_sql);
+
+            // Issue table create sql query
+            NpgsqlCommand pgsqlCommand = new NpgsqlCommand(table_create_sql, conn);
+            pgsqlCommand.ExecuteNonQuery();
+            Console.WriteLine("Client Table Created Successfully...");
```
Comment (tomtupycape) [inline CapeMSSQLFreyrLink/FreyrManager.cs]: if it fails, it will throw an exception and get caught

File: CapeMSSQLFreyrLink/FreyrManager.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            Console.WriteLine("Freyr - Checking Table Existence: " + tableName);
+            NpgsqlCommand pgsqlCommand = new NpgsqlCommand("SELECT EXISTS (SELECT 1 FROM pg_tables WHERE schemaname = '" + ConfigurationManager.AppSettings["freyrClientsSchema"] + "' AND tablename = '" + tableName + "'); ", conn);
+            return (bool)pgsqlCommand.ExecuteScalar();
+        }
+
+        public void createClientTable(string table_name, List<CapeTableColumn> columns)
+        {
+            // Get type mapping
+            Dictionary<string, string>  MSSQL_TO_POSTGRES_DATA_TYPE_MAP = DBConfigHelper.getMSSQL2PostgreSQLDataTypeMap();
```
Comment (tomtupycape) [inline CapeMSSQLFreyrLink/FreyrManager.cs]: improved readability :D

File: CapeMSSQLFreyrLink/CapeSftpManager.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            this.FileSizeStableIterations = 0;
+            // Get version from filename, use two steps for safety
+            // Ex. CapeData_45_v20170410.mdf
+            this.Version = int.Parse(Regex.Match(sftpFile.FullName, @"_\d+_").Groups[0].Value.Replace("_", ""));
+        }
+    }
+
+    class CapeSftpManager
+    {
+        SftpClient sftpClient;
```
Comment (tomtupycape) [inline CapeMSSQLFreyrLink/CapeSftpManager.cs]: ah, should be snake case

File: CapeMSSQLFreyrLink/CapeDBClasses.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            this.Column = column;
+            this.Value_Integer = value;
+        }
+        public override string ToString()
+        {
+            switch (this.Column.Type) {
+                case "varchar":
+                case "nvarchar":
+                    // Handle single quotes
+                    return "'" + this.Value_String.Replace("'", "''") + "'";
```
Comment (tomtupycape) [inline CapeMSSQLFreyrLink/CapeDBClasses.cs]: yup, that's the proper SQL way to escape a single quote

File: CapeMSSQLFreyrLink/Program.cs (hunk 1 above)
Comment (tomtupycape) [inline CapeMSSQLFreyrLink/Program.cs]: Adding a todo, and when I add book upload functionality, I'll revisit

File: CapeMSSQLFreyrLink/DBConfigHelper.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    }
+
+    /* 
+     * == DBConfigColumnTypeMap ==
+     * Handles the dbTableColumnTypeMapping config section
+     */
+    public class DBConfigColumnTypeMapSection : ConfigurationSection
+    {
+        [ConfigurationProperty("", IsRequired = true, IsDefaultCollection = true)]
```
Comment (tomtupycape) [inline CapeMSSQLFreyrLink/DBConfigHelper.cs]: added comments

**Pull request merged**